File: src/main/java/santiagoAndFerdy/vgs/gridScheduler/GridScheduler.java
package santiagoAndFerdy.vgs.gridScheduler;

import santiagoAndFerdy.vgs.discovery.IRepository;
import santiagoAndFerdy.vgs.discovery.Pinger;
import santiagoAndFerdy.vgs.discovery.Status;
import santiagoAndFerdy.vgs.discovery.selector.Selectors;
import santiagoAndFerdy.vgs.messages.*;
import santiagoAndFerdy.vgs.resourceManager.IResourceManager;
import santiagoAndFerdy.vgs.rmi.RmiServer;

import java.rmi.RemoteException;
import java.rmi.server.UnicastRemoteObject;
import java.util.*;

public class GridScheduler extends UnicastRemoteObject implements IGridScheduler {
    private static final long              serialVersionUID = -5694724140595312739L;

    private RmiServer                      rmiServer;
    private int                            id;
    private IRepository<IResourceManager>  rmRepository;
    private IRepository<IGridScheduler>    gsRepository;
    private Pinger<IResourceManager>       resourceManagerPinger;
    private Pinger<IGridScheduler>         gridSchedulerPinger;

    private Map<Integer, Set<WorkRequest>> monitoredJobs;
    private Map<WorkRequest, MonitorRequest> pendingMonitoringRequests;
    private Map<Integer, Set<BackUpRequest>> backUpJobs;
    private Map<WorkRequest, Integer> pendingBackUpRequests;

    private boolean                        running;

    private long                           load;

    public GridScheduler(RmiServer rmiServer, int id, IRepository<IResourceManager> rmRepository, IRepository<IGridScheduler> gsRepository)
            throws RemoteException {
        this.rmiServer = rmiServer;
        this.id = id;
        rmiServer.register(gsRepository.getUrl(id), this);

        this.rmRepository = rmRepository;
        this.gsRepository = gsRepository;
        gridSchedulerPinger = new Pinger(gsRepository);
        resourceManagerPinger = new Pinger(rmRepository);

        start();
    }

    @Override
    public void monitor(MonitorRequest monitorRequest) throws RemoteException {
        if (!running) throw new RemoteException("I am offline");

        System.out.println("[GS\t" + id + "] Received job " + monitorRequest.getWorkRequest().getJob().getJobId() + " to monitor at GS " + id);

        monitoredJobs.get(monitorRequest.getSourceResourceManagerId()).add(monitorRequest.getWorkRequest());

        BackUpRequest backUpRequest = new BackUpRequest(monitorRequest.getWorkRequest(), id, 0);
        backUp(backUpRequest);
        gsRepository.invokeOnEntity((gs, gsId) -> {
            System.out.println("[GS\t" + id + "] Sending backup monitoring request to GS " + gsId + " for job " + monitorRequest.getWorkRequest().getJob().getJobId());
            pendingMonitoringRequests.put(monitorRequest.getWorkRequest(), monitorRequest);
            gs.backUp(backUpRequest);

            return null;
        }, Selectors.invertedWeighedRandom, id);
    }

    @Override
    public void backUp(BackUpRequest backUpRequest) throws RemoteException {
        if (!running) throw new RemoteException("I am offline");
        WorkRequest work = backUpRequest.getWorkRequest();


        if(backUpRequest.getBackUpsRequested() == 0) {
            // I'm the last, send reply directly
            BackUpAck ack = new BackUpAck(work, id);

            int ackToId = backUpRequest.getTrail()[backUpRequest.getTrail().length - 1];
            Optional<IGridScheduler> ackTo = gsRepository.getEntity(ackToId);

            if(ackTo.isPresent()) {
                ackTo.ifPresent(gs -> {
                    try {
                        System.out.println("[GS\t" + id + "] Received backup request from with trail " + Arrays.toString(backUpRequest.getTrail()) + " and " + backUpRequest.getBackUpsRequested());
                        gs.acceptBackUpAck(ack);
                    } catch (RemoteException e) {
                        e.printStackTrace();
                    }
                });
            } else {

            }
        } else {
            // Make more backups
            BackUpRequest newBackUpRequest = backUpRequest.hop(id);
            Optional<BackUpAck> ack =  gsRepository.invokeOnEntity((gs, gsId) -> {
                System.out.println("[GS\t" + id + "] Sending job " + work.getJob().getJobId() + " to GS " + id + " to make more backups.");
                pendingBackUpRequests.put(work, backUpRequest.getTrail()[backUpRequest.getTrail().length - 1]);

                gs.backUp(newBackUpRequest);
                // TODO dubble check

                return null;
            }, Selectors.invertedWeighedRandom, newBackUpRequest.getTrail());
        }
    }

    @Override
    public synchronized void acceptBackUpAck(BackUpAck ack) throws RemoteException {
        WorkRequest work = ack.getWorkRequest();
        System.out.println("[GS\t" + id + "] Send backup ack for job " + work.getJob().getJobId() + " from " + Arrays.toString(ack.getBackUps()) + " to RM");
        Optional<MonitorRequest> maybeMonitorRequest = Optional.ofNullable(pendingMonitoringRequests.remove(ack.getWorkRequest()));
        if(maybeMonitorRequest.isPresent()) {
            maybeMonitorRequest.ifPresent(monitorRequest -> {
                Optional<IResourceManager> maybeRm = rmRepository.getEntity(monitorRequest.getSourceResourceManagerId());

                if(maybeRm.isPresent()) {
                    try {
                        maybeRm.get().monitorAck(ack.prependGridSchedulerList(id));
                    } catch (RemoteException e) {
                        // TODO Reschedule
                        e.printStackTrace();
                    }
                } else {
                    // TODO Reschedule
                }
            });
        } else {
            Optional<Integer> maybeGsToAckId = Optional.ofNullable(pendingBackUpRequests.remove(ack.getWorkRequest()));

            if(maybeGsToAckId.isPresent()) {
                System.out.println("[GS\t" + id + "] Propagate backup ack for job " + work.getJob().getJobId() + " from " + Arrays.toString(ack.getBackUps()));
                maybeGsToAckId.flatMap(gsRepository::getEntity).ifPresent(gsToAck -> {
                    try {
                        gsToAck.acceptBackUpAck(ack.prependGridSchedulerList(id));
                    } catch (RemoteException e) {
                        // can't be reached? Then do nothing. first part of the chain will do recovery.
                        e.printStackTrace();
                    }
                });
            }
        }
    }

    public void registerBackUp(BackUpAck ack) {

    }

    @Override
    public void promote(PromotionRequest promotionRequest) throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

        System.out.println("[GS\t" + id + "] Promoting to primary for job " + promotionRequest.getToBecomePrimaryFor().getJob()
                .getJobId() + " at cluster " + id);
        backUpJobs.get(promotionRequest.getSourceResourceManagerId()).remove(promotionRequest.getToBecomePrimaryFor());
        monitoredJobs.get(promotionRequest.getSourceResourceManagerId()).add(promotionRequest.getToBecomePrimaryFor());

    }

    @Override
    public void offLoad(WorkRequest workRequest) throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

    }

    @Override
    public void releaseMonitored(WorkRequest request) throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

        monitoredJobs.remove(request);
        System.out.println("[GS\t" + id + "] Stop monitoring " + request.getJob().getJobId() + " at cluster " + id);
    }

    @Override
    public void releaseBackUp(WorkRequest workRequest) throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

        backUpJobs.remove(workRequest);
        System.out.println("[GS\t" + id + "] Releasing back-up of workRequest " + workRequest.getJob().getJobId() + " at cluster " + id);
    }

    @Override
    public void start() throws RemoteException {
        running = true;
        monitoredJobs = new HashMap<>();
        pendingMonitoringRequests = new HashMap<>();
        backUpJobs = new HashMap<>();
        pendingBackUpRequests = new HashMap<>();

        for (int rmId : rmRepository.ids()) {
            monitoredJobs.put(rmId, new HashSet<>());
            backUpJobs.put(rmId, new HashSet<>());
            rmRepository.getEntity(rmId).ifPresent(rm -> {
                try {
                    rm.receiveGridSchedulerWakeUpAnnouncement(id);
                } catch (RemoteException e) {
                }
            });
        }

        for (int gsId : gsRepository.idsExcept(id)) {
            gsRepository.getEntity(gsId).ifPresent(gs -> {
                try {
                    gs.receiveGridSchedulerWakeUpAnnouncement(id);
                } catch (RemoteException e) {
                    // Can be offline. that's okay.
                }
            });
        }

        load = 1;

        gridSchedulerPinger.start();
        resourceManagerPinger.start();

        System.out.println("[GS\t" + id + "] Online");
    }

    @Override
    public void shutDown() throws RemoteException {
        running = false;
        monitoredJobs = null;
        backUpJobs = null;

        gridSchedulerPinger.stop();
        resourceManagerPinger.stop();

        System.out.println("[GS\t" + id + "] Offline");
    }

    @Override
    public void receiveResourceManagerWakeUpAnnouncement(int from) throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

        System.out.println("[GS\t" + id + "] RM " + from + " awake");
        rmRepository.setLastKnownStatus(from, Status.ONLINE);
    }

    @Override
    public void receiveGridSchedulerWakeUpAnnouncement(int from) throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

        System.out.println("[GS\t" + id + "] GS " + from + " awake");
        gsRepository.setLastKnownStatus(from, Status.ONLINE);
    }

    @Override
    public long ping() throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

        return load;
    }

    @Override
    public int getId() throws RemoteException {
        if (!running)
            throw new RemoteException("I am offline");

        return id;
    }
}

File: src/main/scala/grid/gridScheduler/GridScheduler.scala
package grid.gridScheduler

import java.rmi.RemoteException
import java.rmi.server.UnicastRemoteObject
import java.util.concurrent.Executors

import com.typesafe.scalalogging.LazyLogging
import grid.cluster.{Pinger, RemoteShutDown}
import grid.discovery.Repository
import grid.discovery.Selector.InvertedRandomWeighedSelector
import grid.messages._
import grid.resourceManager.IResourceManager
import grid.rmi.RmiServer

import collection.mutable._
import scala.concurrent._
import scala.util.Try

class GridScheduler(val id: Int,
                    val rmRepo: Repository[IResourceManager],
                    val gsRepo: Repository[IGridScheduler]) extends UnicastRemoteObject with IGridScheduler with RemoteShutDown with LazyLogging {

  RmiServer.register(this)

  var online = false

  var gsPinger = new Pinger[IGridScheduler](gsRepo)
  var rmPinger = new Pinger[IResourceManager](rmRepo)

  var monitoringForRm: Map[WorkRequest, Int] = null
  var monitoringPerRm: Map[Int, Set[WorkRequest]] = null
  var backUpForRm: Map[WorkRequest, Int] = null
  var backUpPerRm: Map[Int, Set[WorkRequest]] = null
  var backUpForGs: Map[WorkRequest, Int] = null
  var backUpPerGs: Map[Int, Set[WorkRequest]] = null

  var load = 0


  implicit val executionContext: ExecutionContext = ExecutionContext.fromExecutorService(
    Executors.newWorkStealingPool(128)
  )

  start()

  override def start(): Unit = synchronized {
    monitoringForRm = Map()
    monitoringPerRm = Map()
    backUpForRm = Map()
    backUpPerRm = Map()
    backUpForGs = Map()
    backUpPerGs = Map()

    gsRepo.ids().foreach(gsId => {
      monitoringPerRm.put(gsId, Set())
      backUpPerRm.put(gsId, Set())
      backUpPerGs.put(gsId, Set())
    })

    load = 0

    gsPinger.start()
    rmPinger.start()

    online = true

    logger.info(s"[GS\t${id}] Online")
  }

  override def shutDown(): Unit = synchronized {
    monitoringForRm = Map()
    monitoringPerRm = Map()
    backUpForRm = Map()
    backUpPerRm = Map()
    backUpForGs = Map()
    backUpPerGs = Map()

    gsRepo.ids().foreach(gsId => {
      monitoringPerRm.put(gsId, Set())
      backUpPerRm.put(gsId, Set())
      backUpPerGs.put(gsId, Set())
    })

    load = 0

    gsPinger.stop()
    rmPinger.stop()

    online = false

    logger.info(s"[GS\t${id}] Offline")
  }

  override def monitor(req: MonitorRequest): Unit = ifOnline {
    registerMonitor(req.work, req.rmId)
  }

  def registerMonitor(work: WorkRequest, rmId: Int): Unit = synchronized {
    monitoringForRm.put(work, rmId)
    monitoringPerRm(rmId) += work
    logger.info(s"[GS\t${id}] Monitoring job ${work.job.id} on rm ${rmId}")
  }

  def unregisterMonitor(work: WorkRequest): Unit = synchronized {
    Try { monitoringPerRm(monitoringForRm(work)) -= work }
    Try { monitoringForRm -= work }
  }


  override def releaseMonitor(work: WorkRequest): Unit = ifOnline {
    logger.info(s"[GS\t${id}] Releasing monitoring for job ${work.job.id}")
    unregisterMonitor(work)
  }

  override def backUp(req: BackUpRequest): Unit = ifOnline {
    registerBackUp(req.work, req.rmId, req.monitorId)
  }

  def registerBackUp(work: WorkRequest, rmId: Int, monitorId: Int): Unit = synchronized {
    backUpForRm.put(work, rmId)
    backUpPerRm(rmId) += work

    backUpForGs.put(work, monitorId)
    backUpPerGs(monitorId) += work
    logger.info(s"[GS\t${id}] Backing up job ${work.job.id}")
  }

  def unregisterBackUp(work: WorkRequest): Unit = synchronized {
    Try { backUpPerRm(backUpForRm(work)) -= work }
    Try { backUpForRm -= work }

    Try { backUpPerGs(backUpForGs(work)) -= work }
    Try { backUpForGs -= work }
  }

  override def releaseBackUp(work: WorkRequest): Unit = ifOnline {
    logger.info(s"[GS\t${id}] Releasing back up for job ${work.job.id}")
    unregisterBackUp(work)
  }

  override def promote(req: PromoteRequest): Unit = ifOnline {
    logger.info(s"[GS\t${id}] Promoting to primary for job ${req.work.job.id} for rm ${req.rmId}")
    unregisterBackUp(req.work)
    registerMonitor(req.work, req.rmId)
  }

  override def offLoad(req: OffLoadRequest): Unit = ifOnline {
    rmRepo.invokeOnEntity((rm, rmId) => {
      logger.info(s"[GS\t${id}] Offloading job ${req.work.job.id} to rm ${rmId}")
      Future { rm.orderWork(WorkOrder(req.work, id), true) }
    }, InvertedRandomWeighedSelector) foreach {
      case newRmId => registerMonitor(req.work, newRmId)
    }
  }

  rmRepo.onOffline(rmId => {
    val jobsToReschedule = monitoringPerRm(rmId).clone
    jobsToReschedule.foreach(work => {
      logger.info(s"[GS\t${id}] Rescheduling job ${work.job.id}")

      rmRepo.invokeOnEntity((rm, newRmId) => {
        Future { blocking { rm.orderWork(WorkOrder(work, id), false) } }
      }, InvertedRandomWeighedSelector, rmId) foreach {
        case newRmId => synchronized {
          unregisterMonitor(work)
          registerMonitor(work, newRmId)
        }
      }
    })

    val backedUpJobs = backUpPerRm(rmId).clone

    backedUpJobs.foreach(work => {
      // monitor still alive?
      val monitorId = backUpForGs(work)
      logger.info(s"[GS\t${id}] Recovering rm crash for backed up job ${work.job.id} at ${monitorId}")
      if(gsRepo.checkStatus(monitorId)) {
        logger.info(s"[GS\t${id}] Monitor (gs ${monitorId}) still up, release job")
//        releaseBackUp(work)
      } else {
        logger.info(s"[GS\t${id}] Monitor down, promote, reschedule")
        // monitor down => promote, reschedule
        rmRepo.invokeOnEntity((rm, newRmId) => {
          logger.info(s"[GS\t${id}] Rescheduling job ${work.job.id} as monitor")
          Future { blocking { rm.orderWork(WorkOrder(work, id), false) } }
        }, InvertedRandomWeighedSelector) foreach {
          case newRmId => promote(PromoteRequest(work, newRmId))
        }
      }
    })
  })

  gsRepo.onOffline(gsId => {
    val backUpJobs = backUpPerGs(gsId).clone

    backUpJobs.foreach(work => {
      val rmId = backUpForRm(work)

      // check if rm is still running
      if(!rmRepo.checkStatus(rmId)) {
        // no -> monitor and reschedule, otherwise rm will take care of itself
        logger.info(s"[GS\t${id}] Use backup for job ${work.job.id}")
        unregisterBackUp(work)
        rmRepo.invokeOnEntity((rm, newRmId) => {
          Future { rm.orderWork(WorkOrder(work, newRmId), false) }
        }, InvertedRandomWeighedSelector, rmId) foreach {
          case newRmId => registerMonitor(work, newRmId)
        }
      } else {
//        releaseBackUp(work)
      }
    })
  })

  @throws(classOf[RemoteException])
  override def ping(): Long = ifOnline { load }

  override def url: String = gsRepo.url(id)
}
